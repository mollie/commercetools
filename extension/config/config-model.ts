export interface Config {
  port?: number | string;
  mollieApiKey: string;
  ctConfig: {
    projectKey: string;
    clientId: string;
    clientSecret: string;
    authUrl: string;
    host: string;
    scopes?: string[];
  };
}
