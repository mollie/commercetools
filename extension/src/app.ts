import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
