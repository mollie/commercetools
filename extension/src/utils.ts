import { MethodsListParams } from '@mollie/api-client';
/**
 * Generates an ISO string date
 * @returns {String} Returns the current date converted to ISO.
 */
export function createDateNowString() {
  return new Date().toISOString();
}

export function amountMapper(amountPlanned: any): string {
  const { centAmount, fractionDigits } = amountPlanned;
  const divider = Math.pow(10, fractionDigits || 2);
  const mollieAmount = (centAmount / divider).toFixed(2);
  return mollieAmount
}

export function methodListMapper(ctObj: any): MethodsListParams {
  // Generally this shouldn't be needed, but a safety anyway.. eventually could return error here
  if (!ctObj.amountPlanned) {
    return {};
  }
  const mollieAmount = amountMapper(ctObj.amountPlanned);

  const mObject: MethodsListParams = {
    amount: {
      value: mollieAmount,
      currency: ctObj.amountPlanned.currencyCode,
    },
    // Resource is hardcoded, for the time being we only support Orders API
    resource: 'orders',
  };

  if (ctObj.custom?.fields?.paymentMethodsRequest) {
    const parsedMethodsRequest = JSON.parse(ctObj.custom?.fields?.paymentMethodsRequest)
    const { locale, billingCountry, includeWallets, orderLineCategories, issuers, pricing, sequenceType } = parsedMethodsRequest;
    const include = issuers || pricing ? `${issuers ? 'issuers,' : ''}${pricing ? 'pricing' : ''}` : undefined;

    Object.assign(
      mObject,
      locale && { locale: locale },
      include && { include: include },
      includeWallets && { includeWallets: includeWallets },
      billingCountry && { billingCountry: billingCountry },
      sequenceType && { sequenceType: sequenceType },
      orderLineCategories && { orderLineCategories: orderLineCategories },
    );
  }

  return mObject;
}
