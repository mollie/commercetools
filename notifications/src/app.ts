import express, { Application } from 'express';
import handler from '../expressHandler';

const app: Application = express();

app.use(express.urlencoded({ extended: true }));

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use('/', handler);

export default app;
