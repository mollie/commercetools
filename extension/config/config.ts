import { Config } from './config-model';

const isConfigValid = (config: Config): { valid: boolean; message: string } => {
  const { mollie, commercetools } = config;
  let message = '';

  if (!mollie?.apiKey) {
    message = message + 'No Mollie API Key found\n';
  }
  if (!commercetools?.host || !commercetools?.clientId || !commercetools?.clientSecret || !commercetools?.authUrl || !commercetools?.projectKey) {
    message = message + 'Commercetools configuration requires missing required key(s)\n';
  }
  return {
    valid: !message.length,
    message,
  };
};

export function loadConfig(ctMollieConfig: string | undefined) {
  console.info("loadConfig : " + ctMollieConfig);
  try {
    const envConfig = JSON.parse(ctMollieConfig || '');

    const config: Config = {
      ...envConfig,
      service: {
        port: envConfig.service?.port || 3000,
        logLevel: process.env.LOG_LEVEL || envConfig.service?.logLevel || 'info',
        logTransports: envConfig.service?.logTransports || 'terminal',
      },
    };

    const localeRegex = /^[a-z]{2}_[A-Z]{2}$/;

    Object.assign(
      config.service,
      envConfig.service?.webhookUrl && { webhookUrl: envConfig.service.webhookUrl },
      envConfig.service?.redirectUrl && { redirectUrl: envConfig.service.redirectUrl },
      envConfig.service?.locale && envConfig.service.locale.match(localeRegex) && { locale: envConfig.service.locale },
    );
    Object.assign(config.commercetools, envConfig.commercetools?.authentication ? { authentication: envConfig.commercetools.authentication } : { authentication: { isBasicAuth: false } });
    Object.assign(config.commercetools, { enableRetry: envConfig.commercetools.enableRetry ?? true });

    const { valid, message } = isConfigValid(config);

    if (valid) {
      return config;
    } else {
      throw new Error(message);
    }
  } catch (e) {
    console.error(e);
    throw new Error('Commercetools - Mollie Integration configuration is missing or not provided in the valid JSON format');
  }
}

const ctMollieConfig = process.env.CT_MOLLIE_CONFIG;
const config = loadConfig(ctMollieConfig);

export default config;
