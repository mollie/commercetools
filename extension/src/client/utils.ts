import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}
