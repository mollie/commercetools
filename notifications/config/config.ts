import _ from 'lodash';
import { Config } from './config-model';

const isConfigValid = (config: Config): { valid: boolean; message: string } => {
  const { mollie, commercetools } = config;
  let message = '';

  if (!mollie?.apiKey) {
    message = message + 'No Mollie API Key found\n';
  }
  if (!config.commercetools) {
    message = message + 'No Commercetools configuration present\n';
  }
  if (!commercetools?.host || !commercetools?.clientId || !commercetools?.clientSecret || !commercetools?.authUrl || !commercetools?.projectKey) {
    message = message + 'Commercetools configuration requires missing required key(s)\n';
  }
  return {
    valid: !message.length,
    message,
  };
};

export function loadConfig(ctMollieConfig: string | undefined) {
  try {
    // Parse env config, don't allow running without config
    const envConfig = JSON.parse(ctMollieConfig || '');

    // Allow missing parts of config and fill in with defaults
    const config: Config = {
      service: {
        port: envConfig.service?.port || 3001,
        logLevel: process.env.LOG_LEVEL || envConfig.service?.logLevel || 'info',
        logTransports: envConfig.service?.logTransports || 'terminal',
      },
      ...envConfig,
    };

    const { valid, message } = isConfigValid(config);
    if (valid) {
      return config;
    } else {
      throw new Error(message);
    }
  } catch (e) {
    throw new Error('Commercetools - Mollie Integration configuration is incomplete, missing or not provided in the valid JSON format');
  }
}

const ctMollieConfig = process.env.CT_MOLLIE_CONFIG;
const config = loadConfig(ctMollieConfig);

export default config;
