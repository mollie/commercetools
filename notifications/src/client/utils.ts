import fetch from 'node-fetch-commonjs';
import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';
import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth';
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http';
import { createLoggerMiddleware } from '@commercetools/sdk-middleware-logger';
import { createUserAgentMiddleware } from '@commercetools/sdk-middleware-user-agent';
import { createClient } from '@commercetools/sdk-client';
import Logger from '../logger/logger';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-notifications/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): any {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const userAgentMiddleware = createUserAgentMiddleware({
    libraryName: 'MollieCommercetools-notification',
    libraryVersion: version,
  });

  const ctAuthMiddleware = createAuthMiddlewareForClientCredentialsFlow({
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  });

  const ctHttpMiddleWare = createHttpMiddleware({
    host,
    fetch,
  });

  let commercetoolsClient: any;

  if (Logger.level === 'http' || Logger.level === 'verbose' || Logger.level === 'debug') {
    commercetoolsClient = createClient({ middlewares: [userAgentMiddleware, ctAuthMiddleware, ctHttpMiddleWare, createLoggerMiddleware()] });
  } else {
    commercetoolsClient = createClient({ middlewares: [userAgentMiddleware, ctAuthMiddleware, ctHttpMiddleWare] });
  }
  return commercetoolsClient;
}
