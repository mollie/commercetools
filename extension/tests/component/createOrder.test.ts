import nock from 'nock';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import { mocked } from 'ts-jest/utils';
import app from '../../src/app';
import config from '../../config/config';
import Logger from '../../src/logger/logger';

import {
  noCartFoundForGivenPaymentId,
  cartFoundWith2LineItemsForGivenPaymentId,
  cartFoundWith2LineItemsAndOneCustomLineItemForGivenPaymentId,
} from './mockResponses/commercetoolsData/cartResponses.data';
import {
  paymentMethodNotEnabledInProfile,
  amountLowerThanMinimumKlarnaSliceIt,
  orderCreatedWithTwoLinesUsingIdeal,
  orderCreatedWithTwoLineItemsUsingKlarna,
  orderCreatedIncludingDiscountLineUsingIdeal,
} from './mockResponses/mollieData/createOrder.data';

jest.mock('uuid');

describe('Create Order', () => {
  const {
    commercetools: { host, projectKey, authUrl },
  } = config;
  const mockLogDebug = jest.fn();
  const mockLogError = jest.fn();

  const paymentId = 'd75d0b1d-64c5-4c8f-86f6-b9510332e743';
  const baseMockCTPaymentObj: any = {
    resource: {
      obj: {
        id: paymentId,
        paymentMethodInfo: {
          paymentInterface: 'mollie',
        },
        amountPlanned: {
          currencyCode: 'EUR',
          centAmount: 50000,
        },
        custom: {
          fields: {
            createPayment: '{}',
          },
        },
      },
    },
  };
  let authTokenScope: any;
  beforeAll(() => {
    // Ensure consistent uuid and datetime
    jest.spyOn(Date.prototype, 'toISOString').mockImplementation(() => '2021-11-10T14:02:45.858Z');
    mocked(uuid).mockReturnValue('b2bd1698-9923-4704-9729-02db2de495d1');
    // Credentials authentication flow is called first by commercetools client
    authTokenScope = nock(`${authUrl}`).persist().post('/oauth/token').reply(200, {
      access_token: 'vkFuQ6oTwj8_Ye4eiRSsqMeqLYNeQRJi',
      expires_in: 172800, // seconds (2 days)
      scope: 'manage_project:{projectKey}',
      token_type: 'Bearer',
    });
    // Prevent logs from cluttering test output
    Logger.debug = mockLogDebug;
    Logger.error = mockLogError;
  });

  afterEach(() => {
    nock.cleanAll();
  });

  afterAll(() => {
    authTokenScope.persist(false);
  });

  describe('Unhappy Path', () => {
    it('Payment not linked to Cart', async () => {
      const getCartByPaymentIdScope = nock(`${host}/${projectKey}`)
        .get('/carts')
        .query(true) // mock the url regardless of query string
        .reply(200, noCartFoundForGivenPaymentId);

      const mockCTPaymentObj = {
        ...baseMockCTPaymentObj,
      };
      mockCTPaymentObj.resource.obj.paymentMethodInfo['method'] = 'ideal';
      mockCTPaymentObj.resource.obj['transactions'] = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: 'Charge',
          state: 'Initial',
          amount: {
            currencyCode: 'EUR',
            centAmount: 50000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);

      const { status, text } = res;
      expect(status).toBe(400);

      const parsedText = JSON.parse(text);
      const { errors } = parsedText;

      expect(errors).toHaveLength(1);
      expect(errors[0]).toEqual({
        code: 'ObjectNotFound',
        message: `Could not find Cart associated with the payment ${paymentId}.`,
        extensionExtraInfo: {
          originalStatusCode: 404,
        },
      });
      expect(getCartByPaymentIdScope.isDone()).toBeTruthy();
    });

    it('422 from mollie - pay now method', async () => {
      const getCartByPaymentIdScope = nock(`${host}/${projectKey}`)
        .get('/carts')
        .query(true) // mock the url regardless of query string
        .reply(200, cartFoundWith2LineItemsForGivenPaymentId);

      const createOrderFailsAsUnprocessableScope = nock('https://api.mollie.com/v2').post('/orders?embed=payments').reply(422, paymentMethodNotEnabledInProfile);

      const mockCTPaymentObj = {
        ...baseMockCTPaymentObj,
      };
      mockCTPaymentObj.resource.obj.paymentMethodInfo['method'] = 'ideal';
      mockCTPaymentObj.resource.obj['transactions'] = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: 'Charge',
          state: 'Initial',
          amount: {
            currencyCode: 'EUR',
            centAmount: 50000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);

      const { status, text } = res;
      expect(status).toBe(400);

      const parsedText = JSON.parse(text);
      const { errors } = parsedText;

      expect(errors).toHaveLength(1);
      expect(errors[0]).toEqual({
        code: 'SemanticError',
        message: 'The payment method is not enabled in your website profile',
        extensionExtraInfo: {
          originalStatusCode: 422,
          field: 'payment.method',
          title: 'Unprocessable Entity',
          links: {
            documentation: {
              href: 'https://docs.mollie.com/reference/v2/orders-api/create-order',
              type: 'text/html',
            },
          },
        },
      });

      expect(getCartByPaymentIdScope.isDone()).toBeTruthy();
      expect(createOrderFailsAsUnprocessableScope.isDone()).toBeTruthy();
    });

    it('422 from mollie - pay later method', async () => {
      const getCartByPaymentIdScope = nock(`${host}/${projectKey}`)
        .get('/carts')
        .query(true) // mock the url regardless of query string
        .reply(200, cartFoundWith2LineItemsForGivenPaymentId);

      const createOrderFailsAsUnprocessableScope = nock('https://api.mollie.com/v2').post('/orders?embed=payments').reply(422, amountLowerThanMinimumKlarnaSliceIt);

      const mockCTPaymentObj = {
        ...baseMockCTPaymentObj,
      };
      mockCTPaymentObj.resource.obj.paymentMethodInfo['method'] = 'klarnasliceit';
      mockCTPaymentObj.resource.obj['transactions'] = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: 'Authorization',
          state: 'Initial',
          amount: {
            currencyCode: 'EUR',
            centAmount: 50000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);

      const { status, text } = res;
      expect(status).toBe(400);

      const parsedText = JSON.parse(text);
      const { errors } = parsedText;

      expect(errors).toHaveLength(1);
      expect(errors[0]).toEqual({
        code: 'SemanticError',
        message: 'The amount is lower than the minimum',
        extensionExtraInfo: {
          originalStatusCode: 422,
          field: 'payment.amount',
          title: 'Unprocessable Entity',
          links: {
            documentation: {
              href: 'https://docs.mollie.com/reference/v2/orders-api/create-order',
              type: 'text/html',
            },
          },
        },
      });

      expect(getCartByPaymentIdScope.isDone()).toBeTruthy();
      expect(createOrderFailsAsUnprocessableScope.isDone()).toBeTruthy();
    });
  });

  describe('Happy Path', () => {
    afterEach(() => {
      nock.cleanAll();
    });

    it('200 - Order created successfully using pay now method (iDEAL)', async () => {
      const getCartByPaymentIdScope = nock(`${host}/${projectKey}`)
        .get('/carts')
        .query(true) // mock the url regardless of query string
        .reply(200, cartFoundWith2LineItemsForGivenPaymentId);
      const orderCreatedWithPayNowMethodScope = nock('https://api.mollie.com/v2').post('/orders?embed=payments').reply(201, orderCreatedWithTwoLinesUsingIdeal);

      const mockCTPaymentObj = {
        ...baseMockCTPaymentObj,
      };
      mockCTPaymentObj.resource.obj.paymentMethodInfo['method'] = 'ideal';
      mockCTPaymentObj.resource.obj['transactions'] = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: 'Charge',
          state: 'Initial',
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);
      const { status, text } = res;
      expect(status).toBe(201);

      const parsedActions = JSON.parse(text);
      const { actions } = parsedActions;
      expect(actions).toHaveLength(6);

      // Ensure the interface interaction contains the checkout url
      const interfaceInteractionAction = actions.find((action: any) => action.action === 'addInterfaceInteraction');
      expect(JSON.parse(interfaceInteractionAction.fields.response)).toEqual({
        mollieOrderId: 'ord_8xnw8a',
        checkoutUrl: 'https://www.mollie.com/checkout/order/8xnw8a',
        transactionId: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
      });
      // Snapshot all update actions
      actions.forEach((action: any) => {
        expect(action).toMatchSnapshot();
      });
      expect(getCartByPaymentIdScope.isDone()).toBeTruthy();
      expect(orderCreatedWithPayNowMethodScope.isDone()).toBeTruthy();
    });

    it('200 - Order created successfully using pay later method (Klarnapaylater)', async () => {
      const getCartByPaymentIdScope = nock(`${host}/${projectKey}`)
        .get('/carts')
        .query(true) // mock the url regardless of query string
        .reply(200, cartFoundWith2LineItemsForGivenPaymentId);
      const orderCreatedWithPayLaterMethodScope = nock('https://api.mollie.com/v2').post('/orders?embed=payments').reply(201, orderCreatedWithTwoLineItemsUsingKlarna);
      const mockCTPaymentObj = {
        ...baseMockCTPaymentObj,
      };
      mockCTPaymentObj.resource.obj.paymentMethodInfo['method'] = 'klarnapaylater';
      mockCTPaymentObj.resource.obj['transactions'] = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: 'Authorization',
          state: 'Initial',
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);
      const { status, text } = res;
      expect(status).toBe(201);

      const parsedActions = JSON.parse(text);
      const { actions } = parsedActions;
      expect(actions).toHaveLength(6);

      // Ensure the interface interaction contains the checkout url
      const interfaceInteractionAction = actions.find((action: any) => action.action === 'addInterfaceInteraction');
      expect(JSON.parse(interfaceInteractionAction.fields.response)).toEqual({
        mollieOrderId: 'ord_l2idwq',
        checkoutUrl: 'https://www.mollie.com/checkout/order/l2idwq',
        transactionId: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
      });
      // Snapshot all update actions
      actions.forEach((action: any) => {
        expect(action).toMatchSnapshot();
      });

      expect(getCartByPaymentIdScope.isDone()).toBeTruthy();
      expect(orderCreatedWithPayLaterMethodScope.isDone()).toBeTruthy();
    });

    it('200 - Order created from cart with Line Items and Custom Line Items', async () => {
      const getCartByPaymentIdScope = nock(`${host}/${projectKey}`)
        .get('/carts')
        .query(true) // mock the url regardless of query string
        .reply(200, cartFoundWith2LineItemsAndOneCustomLineItemForGivenPaymentId);
      const orderCreatedWithDiscountLineScope = nock('https://api.mollie.com/v2').post('/orders?embed=payments').reply(201, orderCreatedIncludingDiscountLineUsingIdeal);

      const mockCTPaymentObj = {
        ...baseMockCTPaymentObj,
      };
      mockCTPaymentObj.resource.obj.id = '990d9419-62c2-44e5-91d4-8cb9e5cc6518';
      mockCTPaymentObj.resource.obj.paymentMethodInfo['method'] = 'ideal';
      mockCTPaymentObj.resource.obj['transactions'] = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: 'Charge',
          state: 'Initial',
          amount: {
            currencyCode: 'EUR',
            centAmount: 88500,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);
      const { status, text } = res;
      expect(status).toBe(200);

      const parsedActions = JSON.parse(text);
      const { actions } = parsedActions;
      expect(actions).toHaveLength(6);

      // Ensure the interface interaction contains the checkout url
      const interfaceInteractionAction = actions.find((action: any) => action.action === 'addInterfaceInteraction');
      expect(JSON.parse(interfaceInteractionAction.fields.response)).toEqual({
        mollieOrderId: 'ord_ca1j7q',
        checkoutUrl: 'https://www.mollie.com/checkout/order/ca1j7q',
        transactionId: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
      });
      // Snapshot all update actions
      actions.forEach((action: any) => {
        expect(action).toMatchSnapshot();
      });
      expect(getCartByPaymentIdScope.isDone()).toBeTruthy();
      expect(orderCreatedWithDiscountLineScope.isDone()).toBeTruthy();
    });
  });
});
