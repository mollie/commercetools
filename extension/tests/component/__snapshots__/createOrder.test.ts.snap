// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Order Happy Path Should return 201 when mollie order created from cart with both Line Items and Custom Line Items 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "createOrder",
    "createdAt": "2021-12-21T07:33:07+00:00",
    "id": "b2bd1698-9923-4704-9729-02db2de495d1",
    "request": "{\\"cartId\\":\\"84e5f274-c413-404e-b938-ffb45ce63f61\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\",\\"createPayment\\":{}}",
    "response": "{\\"mollieOrderId\\":\\"ord_ca1j7q\\",\\"checkoutUrl\\":\\"https://www.mollie.com/checkout/order/ca1j7q\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\"}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created from cart with both Line Items and Custom Line Items 2`] = `
Object {
  "action": "setStatusInterfaceText",
  "interfaceText": "created",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created from cart with both Line Items and Custom Line Items 3`] = `
Object {
  "action": "setKey",
  "key": "ord_ca1j7q",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created from cart with both Line Items and Custom Line Items 4`] = `
Object {
  "action": "changeTransactionState",
  "state": "Pending",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created from cart with both Line Items and Custom Line Items 5`] = `
Object {
  "action": "changeTransactionInteractionId",
  "interactionId": "tr_D6V5dfjWfz",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created from cart with both Line Items and Custom Line Items 6`] = `
Object {
  "action": "changeTransactionTimestamp",
  "timestamp": "2021-12-21T07:33:07+00:00",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay later method (Klarnapaylater) 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "createOrder",
    "createdAt": "2021-12-20T11:17:23+00:00",
    "id": "b2bd1698-9923-4704-9729-02db2de495d1",
    "request": "{\\"cartId\\":\\"08f21547-92c8-4519-9fd8-84dc05827f0f\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\",\\"createPayment\\":{}}",
    "response": "{\\"mollieOrderId\\":\\"ord_l2idwq\\",\\"checkoutUrl\\":\\"https://www.mollie.com/checkout/order/l2idwq\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\"}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay later method (Klarnapaylater) 2`] = `
Object {
  "action": "setStatusInterfaceText",
  "interfaceText": "created",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay later method (Klarnapaylater) 3`] = `
Object {
  "action": "setKey",
  "key": "ord_l2idwq",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay later method (Klarnapaylater) 4`] = `
Object {
  "action": "changeTransactionState",
  "state": "Pending",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay later method (Klarnapaylater) 5`] = `
Object {
  "action": "changeTransactionInteractionId",
  "interactionId": "tr_FCc42uPfJx",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay later method (Klarnapaylater) 6`] = `
Object {
  "action": "changeTransactionTimestamp",
  "timestamp": "2021-12-20T11:17:23+00:00",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay now method (iDEAL) 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "createOrder",
    "createdAt": "2021-12-20T10:15:15+00:00",
    "id": "b2bd1698-9923-4704-9729-02db2de495d1",
    "request": "{\\"cartId\\":\\"08f21547-92c8-4519-9fd8-84dc05827f0f\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\",\\"createPayment\\":{}}",
    "response": "{\\"mollieOrderId\\":\\"ord_8xnw8a\\",\\"checkoutUrl\\":\\"https://www.mollie.com/checkout/order/8xnw8a\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\"}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay now method (iDEAL) 2`] = `
Object {
  "action": "setStatusInterfaceText",
  "interfaceText": "created",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay now method (iDEAL) 3`] = `
Object {
  "action": "setKey",
  "key": "ord_8xnw8a",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay now method (iDEAL) 4`] = `
Object {
  "action": "changeTransactionState",
  "state": "Pending",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay now method (iDEAL) 5`] = `
Object {
  "action": "changeTransactionInteractionId",
  "interactionId": "tr_J5R48EP58J",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Happy Path Should return 201 when mollie order created successfully using pay now method (iDEAL) 6`] = `
Object {
  "action": "changeTransactionTimestamp",
  "timestamp": "2021-12-20T10:15:15+00:00",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;
